### ISSUE TITLE
Conversion to Linalg on Tensors seems (very) incomplete

### ISSUE THREAD

**Issue by dpotop** [google/iree#5186]
Following a [discussion 10 days ago](https://github.com/google/iree/issues/5072), I learned that ConvertHLOToLinalgOnBuffersPass is to be deprecated and replaced with lowering HLO to linalg-on-tensors. I have also learned that this new lowering code has already been upstreamed to the MHLO repository.

I have therefore tried to use the last version of the MHLO transformations, available as ```tf-opt --hlo-legalize-to-linalg``` (after pulling the most recent version of TF, which includes MHLO).

On my RESNET50 example, the pass cannot transform two fundamental operations:
* mhlo.convolution, if there is padding
* mhlo.constant, if the constant is not a scalar

I have simplified the input to the following 3 functions:
```
  func @test_conv(%750: tensor<1x7x7x512xf32>,%247:tensor<3x3x512x512xf32>) -> (tensor<1x5x5x512xf32>) {
    %751 = "tf.Conv2D"(%750, %247) {
	     padding = "VALID",
	     strides = [1, 1, 1, 1]
	     } : (tensor<1x7x7x512xf32>, tensor<3x3x512x512xf32>) -> tensor<1x5x5x512xf32>
    return %751: tensor<1x5x5x512xf32>
  }
  
  func @test_conv_pad(%750: tensor<1x7x7x512xf32>,%247:tensor<3x3x512x512xf32>) -> (tensor<1x7x7x512xf32>) {
    %751 = "tf.Conv2D"(%750, %247) {
             dilations = [1, 1, 1, 1],
	     padding = "SAME",
	     strides = [1, 1, 1, 1]
	     } : (tensor<1x7x7x512xf32>, tensor<3x3x512x512xf32>) -> tensor<1x7x7x512xf32>
    return %751: tensor<1x7x7x512xf32>
  }

  func @test_const() -> (tensor<f32>,tensor<2xf32>,tensor<3x3x512x512xf32>) {
    %30 = "tf.Const"() {value = dense<1.0> : tensor<f32>} : () -> tensor<f32>
    %31 = "tf.Const"() {value = dense<[1.0,2.0]> : tensor<2xf32>} : () -> tensor<2xf32>
    %247 = "tf.Const"() {value = dense<1.0> : tensor<3x3x512x512xf32>} : () -> tensor<3x3x512x512xf32>
    return %30,%31,%247: tensor<f32>,tensor<2xf32>,tensor<3x3x512x512xf32>
  }
```
The compilation command is:
```tf-opt     --tf-to-hlo-pipeline     --mhlo-test-unfuse-batch-norm     --hlo-legalize-to-linalg conv.mlir```
 
Note in the output that the first convolution, without padding, is correctly translated, whereas the second is left untranslated at MHLO level. Maybe there is another lowering option allowing the conversion in this case?

For constants (in the last function) it's even worse - this limitation did not previously exist in the IREE pipeline, it was introduced now. In my example, only the first mhlo.constant is correctly translated into a standard constant. The other two remain at mhlo.constant level.

I'm on TF commit ac5f2a1f0db1ba1e585559ad4e593fdb3e73713a (I don't know how this is mirrored on the MHLO standalone repo, but a priori it's mirrored).

ADDENDUM: while digging deeper, I found the solution to the first issue, but this one remains inside IREE: it's the option ```-iree-flow-extract-pad-from-conv```. I guess this one should be moved to MHLO, too, and included in the pipeline, because it completes the pass. The mhlo constant translation issue, on the other hand, I don't yet understand.

**Comment by hanhanW** [comment]
Convolution with padding attributes makes Linalg transforms harder, so we extract it to be another operation in IREE. Many transforms in Linalg do not work with operations that has padding attribute.

The [HLOToHLOPreprocessing](https://github.com/google/iree/blob/main/iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp) pass is used to make MHLO ops being IREE-friendly. There are other operations that we want to tweak a bit before jumping into IREE pipeline. If you are relying on them, you're going to want to move these patterns/passes upstream. You can upstream some of them to MHLO repo, like `mhlo::PopulateUnfuseBatchNormPatterns`, `mhlo::PopulateComplexLoweringPatterns`, and `mhlo::PopulateGatherToTorchIndexSelectPatterns`, etc. And integrate them in your project.

I think different projects has different ways to handle constants. In IREE, we inline splat constants and [small constants](https://github.com/google/iree/issues/4897) to dispatch regions. Large constants will be flow.variables or something at flow level. (flow is a dialect defined by IREE). There is a pattern in `iree-codegen-hlo-to-linalg-on-tensors` that lowers all kinds of mhlo.constants to std.constants, but I did not upstream it to MHLO repo yet. The logic of the pattern in MHLO only handle scalar cases, but maybe it is okay to upstream it.

https://github.com/google/iree/blob/96e9c4946399f2a0df6794252c77506e9ac1c9a5/iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp#L231-L239

**Comment by hanhanW** [comment]
FYI, you can run resnet model with IREE. We have a test file with fake weight: https://github.com/google/iree/blob/main/iree/test/e2e/models/resnet_fake_weights.mlir

And there are some integration tests for ResNet50 in integrations/tensorflow/e2e/keras/applications/

It looks like there is no action item on IREE side, so I'm going to close the issue.
